### PR TITLE
HIVE-23629: Enforce clean findbugs in PRs

### PR DESCRIPTION
**Pull request by mustafaiman**
Change-Id: Ided89254e2464cf9a6f5ebfdce5c1f222988d18e

### CODE REVIEW

**Comment by pgaref** [comment]
Seems to be doing the job, check here: http://130.211.9.232/blue/organizations/jenkins/panos-precommit/detail/panos-precommit/3/pipeline

I was just wondering if we should push the results back to JIRA in the form of a report or something..

**Comment by mustafaiman** [comment]
@pgaref The violations are available in pipeline logs. I have mixed feelings about pushing back the results to associated Jira or to the PR. Previously these reports were in the form of warning. It makes sense to push it back to JIra then, so people see them before merging changes. Now, you cannot even get a test run without succeeding the report. So if you have a green test run, you know that you did not introduce a new violation. If the pipeline fails because of a new violation, it is easy enough to see the violations in pipeline logs. Also the report becomes stale as soon as you push new changes. Personally, I really dont like the clutter stale reports create on Jiras. Anyway, I think it is more important to introduce this check as soon as possible. Reporting back can be introduced later if people really miss it.

**Comment by pgaref** [comment]
> @pgaref The violations are available in pipeline logs. I have mixed feelings about pushing back the results to associated Jira or to the PR. Previously these reports were in the form of warning. It makes sense to push it back to JIra then, so people see them before merging changes. Now, you cannot even get a test run without succeeding the report. So if you have a green test run, you know that you did not introduce a new violation. If the pipeline fails because of a new violation, it is easy enough to see the violations in pipeline logs. Also the report becomes stale as soon as you push new changes. Personally, I really dont like the clutter stale reports create on Jiras. Anyway, I think it is more important to introduce this check as soon as possible. Reporting back can be introduced later if people really miss it.

Agreed, lets introduce the check asap -- failing would also save us testing resources as discussed. Reporting can be done as a second step.

File: Jenkinsfile [hunk 1]
```diff
@@ -149,6 +149,13 @@ jobWrappers {
       stage('Checkout') {
         checkout scm
       }
+      stage('Prechecks') {
+        def findbugsProjects = [
+            ":hive-shims-aggregator",
+            ":hive-shims-common"
+        ]
+        buildHive("-Pfindbugs -pl " + findbugsProjects.join(",") + " -am compile findbugs:check")
```
Comment (kgyrtkirk) [inline Jenkinsfile:157]: I think at some point we can put this after the main "compile" step - so that it doesn't need "compile"

**Comment by kgyrtkirk** [comment]
> > @pgaref The violations are available in pipeline logs. I have mixed feelings about pushing back the results to associated Jira or to the PR. Previously these reports were in the form of warning. It makes sense to push it back to JIra then, so people see them before merging changes. Now, you cannot even get a test run without succeeding the report. So if you have a green test run, you know that you did not introduce a new violation. If the pipeline fails because of a new violation, it is easy enough to see the violations in pipeline logs. Also the report becomes stale as soon as you push new changes. Personally, I really dont like the clutter stale reports create on Jiras. Anyway, I think it is more important to introduce this check as soon as possible. Reporting back can be introduced later if people really miss it.
> 
> Agreed, lets introduce the check asap -- failing would also save us testing resources as discussed. Reporting can be done as a next step.

another approach which could be considered: publish the findbugs report in the build - I think it will be there even if the build fails
some of this might work:
https://www.jenkins.io/doc/pipeline/steps/htmlpublisher/
https://www.jenkins.io/blog/2017/02/10/declarative-html-publisher/
I think it could be published in a "finally" block somewhere around compile...

File: shims/common/src/main/java/org/apache/hadoop/hive/shims/CombineHiveKey.java [hunk 2]
```diff
@@ -51,4 +52,14 @@ public int compareTo(Object w) {
...
   }
+
+  @Override
+  public int hashCode() {
+    return super.hashCode();
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    return super.equals(obj);
```
Comment (kgyrtkirk) [inline shims/common/src/main/java/org/apache/hadoop/hive/shims/CombineHiveKey.java:63]: this equals method ignores the `key` field (and hashcode too)

**Comment by mustafaiman** [comment]
> another approach which could be considered: publish the findbugs report in the build - I think it will be there even if the build fails
> some of this might work:
> https://www.jenkins.io/doc/pipeline/steps/htmlpublisher/
> https://www.jenkins.io/blog/2017/02/10/declarative-html-publisher/
> I think it could be published in a "finally" block somewhere around compile...

If we are aiming to enforce clean findbugs in every pull request, I dont see much benefit in having a findbugs report. It is going to be empty all the time apart from the bugs you might introduce in your patch.

**Comment by kgyrtkirk** [comment]
> If we are aiming to enforce clean findbugs in every pull request, I dont see much benefit in having a findbugs report. It is going to be empty all the time apart from the bugs you might introduce in your patch.

I was thinking that it could be used to publish the errors (if there are any) - it's easier to view it on the jenkins... but it can be skipped...

**Pull request merged**